map_view/display-device-location-with-nmea-data-sources/src/main/java/com/esri/samples/display_device_location_with_nmea_data_sources/DisplayDeviceLocationWithNmeaDataSourcesController.java

```diff
@@ -0,0 +1,205 @@
+   */
+  @FXML
+  private void start () {
+
+    // initialize the location data source and prepare to begin receiving location updates when data is pushed. As
+    // updates are received, they will be displayed on the map
+    nmeaLocationDataSource.startAsync();
+
+    // prepare the mock data NMEA sentences
+    File simulatedNmeaDataFile = new File(System.getProperty("data.dir"), "./samples-data/redlands/Redlands.nmea");
+    if (simulatedNmeaDataFile.exists()) {
+
+      try {
+        // read the nmea file contents using a buffered reader and store the mock data sentences in a list
+        BufferedReader bufferedReader = new BufferedReader(new FileReader(simulatedNmeaDataFile.getPath()));
+        // add carriage return for NMEA location data source parser
+        List<String> nmeaSentences = bufferedReader.lines().map(nmeaSentence -> nmeaSentence + "\n").collect(Collectors.toList());
+
+        // close the stream and release resources
+        bufferedReader.close();
+
+        // add a satellite changed listener to the NMEA location data source and display satellite information on the app
+        setupSatelliteChangedListener();
+
+        // create a new timeline to push the mock data NMEA sentences into the data source every 250 ms
+        timeline = new Timeline();
+        timeline.setCycleCount(-1); // loop count
+        timeline.getKeyFrames().add(new KeyFrame(Duration.millis(250), event -> {
+
+          // note: you can also use real-time NMEA sentences obtained via a GPS dongle
+          nmeaLocationDataSource.pushData(nmeaSentences.get(count++).getBytes(StandardCharsets.UTF_8)); // post increment step
+          // reset the count after the last data point is reached
+          if (count == nmeaSentences.size()) count = 0;
+
+        }));
+
+        // start the timeline
+        timeline.play();
+
+        startButton.setDisable(true);
+        stopButton.setDisable(false);
+
+      } catch (Exception e) {
+        new Alert(Alert.AlertType.ERROR, e.getCause().getMessage()).show();
+        e.printStackTrace();
+      }
+
+    } else {
+      new Alert(Alert.AlertType.ERROR, "File not found").show();
+    }
+
+  }
+
+  /**
+   * Obtains NMEA satellite information from the NMEA location data source, and displays satellite information on the app.
+   */
+  private void setupSatelliteChangedListener() {
+
+    HashSet<Integer> uniqueSatelliteIds = new HashSet<>();
+
+    nmeaLocationDataSource.addSatellitesChangedListener(satellitesChangedEvent -> {
+
+      // get satellite information from the NMEA location data source every time the satellites change
+      nmeaSatelliteInfo = satellitesChangedEvent.getSatelliteInfos();
+      // set the text of the satellite count label
+      satelliteCount.setText("Satellite count: " + nmeaSatelliteInfo.size());
+
+      for (NmeaSatelliteInfo satInfo : nmeaSatelliteInfo) {
+        // collect unique satellite ids
+        uniqueSatelliteIds.add(satInfo.getId());
```

Comment:
If the first event has IDs 1, 2, 3 and the next 3, 4, 5 doesn't `uniqueSatelliteIds` end up with 1, 2, 3, 4, 5? If so is that what you wanted to do - show every satellite ID that was ever in a message or are you wanting to show just values from the most recent message? In other words should `uniqueSatelliteIds` be cleared each time?

If your simulated data only has the same IDs in each message you would never notice but if you have variable IDs maybe it is an issue.

Reply:
the sample design for this one is to cumulatively collect the satellite info since we don't have that many messages to go through: the sample exists to show the relevant API that a user could change as appropriate to configure for their own app in an easy to understand why.